File: utils/excel_import.py
"""Excel 批量导入 — 逐行校验 + 错误报告"""
import io
from datetime import datetime

from openpyxl import load_workbook

from database import get_db
from utils.validators import validate_id_number, validate_birth_date_match, validate_date_format, parse_date_input
from utils.helpers import normalize_residence, detect_surname_split


def parse_import_file(file_stream, operator: str = "admin") -> dict:
    """
    解析上传的 Excel 文件，逐行校验，返回导入结果。
    operator 由当前登录会话自动传入（与其它录入入口一致，不需在表格中填写）。
    返回: { total, success, errors: [{row, field, message}], imported_ids: [...] }
    """
    wb = load_workbook(file_stream, read_only=True, data_only=True)
    ws = wb.active
    rows = list(ws.iter_rows(min_row=2, values_only=True))  # 跳过表头

    result = {"total": len(rows), "success": 0, "errors": [], "imported_ids": []}
    db = get_db()

    for i, row in enumerate(rows, start=2):  # 第2行开始
        # 跳过完全空行
        if not row or all(v is None or str(v).strip() == "" for v in row):
            result["total"] -= 1
            continue

        try:
            data = _parse_row(row)
        except Exception as e:
            result["errors"].append({"row": i, "field": "—", "message": f"数据解析失败: {e}"})
            continue

        row_errors = _validate_import_row(data, db)
        if row_errors:
            for err in row_errors:
                result["errors"].append({"row": i, "field": err[0], "message": err[1]})
            continue

        try:
            # 写入 personnel_info
            db.execute(
                "INSERT INTO personnel_info (unit, department, name, gender, birth_date, "
                "id_number, work_start_date, education, degree, title, rank, political_status, "
                "party_join_date, position, operator) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                (
                    data["unit"], data["department"], data["name"], data["gender"],
                    data["birth_date"], data["id_number"], data["work_start_date"], data["education_code"],
                    data["degree_code"], data["title_code"], data["rank_code"],
                    data["political_status"], data["party_join_date"],
                    data["position"], operator,
                ),
            )
            info_id = db.execute("SELECT last_insert_rowid()").fetchone()[0]

            # 写入 personnel_filing
            surname, given_name = detect_surname_split(data["name"])
            db.execute(
                "INSERT INTO personnel_filing (personnel_info_id, surname, given_name, gender, "
                "birth_date, id_number, residence, political_status, work_unit, "
                "position_or_title, supervisor_unit, tag, informed, remarks, operator) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                (
                    info_id, surname, given_name, data["gender"],
                    data["birth_date"], data["id_number"], normalize_residence(data.get("residence", "")),
                    data["political_status"], data["unit"], data.get("position_or_title", ""),
                    data.get("supervisor_unit", "人事处"), "新增", data.get("informed", "是"),
                    data.get("remarks", ""), operator,
                ),
            )

            result["success"] += 1
            result["imported_ids"].append(info_id)
        except Exception as e:
            result["errors"].append({"row": i, "field": "—", "message": f"数据库写入失败: {e}"})

    db.commit()
    return result


def generate_import_template() -> io.BytesIO:
    """生成导入模板 Excel 文件"""
    from openpyxl import Workbook
    from openpyxl.styles import Font, PatternFill

    wb = Workbook()
    ws = wb.active
    ws.title = "备案人员导入模板"

    headers = [
        "单位", "部门", "姓名", "性别", "出生日期", "参加工作日期",
        "身份证号", "户口所在地", "政治面貌", "职务（级）或职称",
        "人事主管单位", "学历", "学位", "职称", "职级",
        "入党日期", "职务（岗位名称）", "标记", "已告知本人", "备注",
    ]
    header_font = Font(bold=True, color="FFFFFF")
    header_fill = PatternFill(start_color="3A5A7C", end_color="3A5A7C", fill_type="solid")

    for col, h in enumerate(headers, 1):
        cell = ws.cell(row=1, column=col, value=h)
        cell.font = header_font
        cell.fill = header_fill

    # 示例行
    example = [
        "XX单位", "XX部门", "张三", "男", "19800103", "20000701",
        "330102198001031230", "浙江杭州市西湖区", "中共党员", "处级",
        "人事处", "大学本科", "学士", "副高", "处级",
        "20050701", "处长", "新增", "是", "",
    ]
    for col, val in enumerate(example, 1):
        ws.cell(row=2, column=col, value=val)

    # 列宽
    col_widths = [18, 14, 10, 6, 12, 12, 20, 22, 14, 18, 14, 12, 10, 10, 10, 12, 18, 8, 12, 20]
    for i, w in enumerate(col_widths, 1):
        from openpyxl.utils import get_column_letter
        ws.column_dimensions[get_column_letter(i)].width = w

    output = io.BytesIO()
    wb.save(output)
    output.seek(0)
    return output


# ---------------------------------------------------------------------------
def _parse_row(row) -> dict:
    """将Excel行转为字典"""
    fields = [
        "unit", "department", "name", "gender", "birth_date", "work_start_date",
        "id_number", "residence", "political_status", "position_or_title",
        "supervisor_unit", "education_code", "degree_code", "title_code", "rank_code",
        "party_join_date", "position", "tag", "informed", "remarks",
    ]
    data = {}
    for i, field in enumerate(fields):
        val = row[i] if i < len(row) else None
        data[field] = str(val).strip() if val is not None else ""
    # 日期清洗
    data["birth_date"] = parse_date_input(data["birth_date"])
    data["work_start_date"] = parse_date_input(data["work_start_date"])
    data["party_join_date"] = parse_date_input(data["party_join_date"])
    return data


def _validate_import_row(data: dict, db) -> list[tuple[str, str]]:
    """返回 [(字段名, 错误信息), ...]"""
    errors = []
    required = [
        ("unit", "单位"), ("department", "部门"), ("name", "姓名"),
        ("gender", "性别"), ("birth_date", "出生日期"), ("id_number", "身份证号"),
        ("political_status", "政治面貌"), ("position", "职务（岗位名称）"),
    ]
    for field, label in required:
        if not data.get(field):
            errors.append((label, "必填项为空"))

    if data.get("birth_date"):
        ok, msg = validate_date_format(data["birth_date"])
        if not ok:
            errors.append(("出生日期", msg))

    if data.get("id_number"):
        ok, msg = validate_id_number(data["id_number"])
        if not ok:
            errors.append(("身份证号", msg))
        elif data.get("birth_date"):
            ok2, msg2 = validate_birth_date_match(data["id_number"], data["birth_date"])
            if not ok2:
                errors.append(("出生日期/身份证号", msg2))

    # 检查重复
    if data.get("id_number"):
        dup = db.execute(
            "SELECT id FROM personnel_filing WHERE id_number = ? AND status = 'active'",
            (data["id_number"],),
        ).fetchone()
        if dup:
            errors.append(("身份证号", "系统中已存在有效备案记录"))

    return errors

File: potms-dotnet/src/POTMS/Services/ExcelImport.cs
using System.Data;
using Dapper;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace POTMS.Services;

/// <summary>Excel 批量导入 —— 逐行校验 + 错误报告。
/// 列顺序与导入模板一致（见 Columns），操作人由登录会话自动写入，不在表格中填写。</summary>
public static class ExcelImport
{
    /// <summary>模板列顺序（与 Python 版 utils/excel_import.py 的 fields 完全一致）。</summary>
    public static readonly (string Field, string Header)[] Columns =
    {
        ("unit", "单位*"), ("department", "部门*"), ("name", "姓名*"), ("gender", "性别*"),
        ("birth_date", "出生日期*(YYYYMMDD)"), ("work_start_date", "参加工作日期"),
        ("id_number", "身份证号*"), ("residence", "户口所在地"), ("political_status", "政治面貌*"),
        ("position_or_title", "职务（级）或职称"), ("supervisor_unit", "人事主管单位"),
        ("education_code", "学历代码"), ("degree_code", "学位代码"), ("title_code", "职称代码"),
        ("rank_code", "职级代码"), ("party_join_date", "入党日期"), ("position", "职务（岗位名称）*"),
        ("tag", "标记"), ("informed", "已告知本人"), ("remarks", "备注"),
    };

    public record RowError(int Row, string Field, string Message);
    public record Result(int Total, int Success, List<RowError> Errors);

    public static Result Parse(IDbConnection cn, Stream stream, string @operator, string? ip)
    {
        var errors = new List<RowError>();
        var rows = ReadRows(stream);
        var total = 0;
        var success = 0;

        foreach (var (rowNumber, cells) in rows)
        {
            if (cells.All(string.IsNullOrWhiteSpace)) continue;   // 跳过完全空行
            total++;

            var data = new Dictionary<string, string?>();
            for (var i = 0; i < Columns.Length; i++)
                data[Columns[i].Field] = i < cells.Count ? (cells[i] ?? "").Trim() : "";
            foreach (var k in new[] { "birth_date", "work_start_date", "party_join_date" })
                data[k] = Validators.ParseDateInput(data[k]);

            var rowErrors = Validate(cn, data);
            if (rowErrors.Count > 0)
            {
                errors.AddRange(rowErrors.Select(e => new RowError(rowNumber, e.Field, e.Message)));
                continue;
            }

            cn.Execute(
                "INSERT INTO personnel_info (unit, department, name, gender, birth_date, id_number, " +
                "work_start_date, education, degree, title, rank, political_status, party_join_date, " +
                "position, operator) VALUES (@unit, @dept, @name, @gender, @birth, @idn, @ws, @edu, " +
                "@deg, @title, @rank, @pol, @pjd, @pos, @op)",
                new
                {
                    unit = data["unit"], dept = data["department"], name = data["name"],
                    gender = data["gender"], birth = data["birth_date"], idn = data["id_number"],
                    ws = data["work_start_date"], edu = data["education_code"], deg = data["degree_code"],
                    title = data["title_code"], rank = data["rank_code"], pol = data["political_status"],
                    pjd = data["party_join_date"], pos = data["position"], op = @operator,
                });
            var infoId = cn.ExecuteScalar<long>("SELECT last_insert_rowid()");

            var (surname, givenName) = Helpers.DetectSurnameSplit(data["name"] ?? "");
            cn.Execute(
                "INSERT INTO personnel_filing (personnel_info_id, surname, given_name, gender, birth_date, " +
                "id_number, residence, political_status, work_unit, position_or_title, supervisor_unit, " +
                "tag, informed, remarks, operator) VALUES (@iid, @sn, @gn, @gender, @birth, @idn, @res, " +
                "@pol, @unit, @pot, @sup, '新增', @informed, @remarks, @op)",
                new
                {
                    iid = infoId, sn = surname, gn = givenName, gender = data["gender"],
                    birth = data["birth_date"], idn = data["id_number"],
                    res = Helpers.NormalizeResidence(data["residence"]),
                    pol = data["political_status"], unit = data["unit"],
                    pot = string.IsNullOrEmpty(data["position_or_title"]) ? data["position"] : data["position_or_title"],
                    sup = string.IsNullOrEmpty(data["supervisor_unit"]) ? "人事处" : data["supervisor_unit"],
                    informed = string.IsNullOrEmpty(data["informed"]) ? "是" : data["informed"],
                    remarks = data["remarks"], op = @operator,
                });
            success++;
        }

        Helpers.LogAction(cn, @operator, ip, "import", "batch",
            detail: $"total={total}, success={success}, errors={errors.Count}");
        return new Result(total, success, errors);
    }

    private static List<(string Field, string Message)> Validate(IDbConnection cn, Dictionary<string, string?> d)
    {
        var errs = new List<(string, string)>();
        foreach (var (field, label) in new[]
                 {
                     ("unit", "单位"), ("department", "部门"), ("name", "姓名"), ("gender", "性别"),
                     ("birth_date", "出生日期"), ("id_number", "身份证号"),
                     ("political_status", "政治面貌"), ("position", "职务（岗位名称）"),
                 })
            if (string.IsNullOrEmpty(d[field])) errs.Add((label, "必填项为空"));

        if (!string.IsNullOrEmpty(d["birth_date"]))
        {
            var (ok, msg) = Validators.ValidateDateFormat(d["birth_date"]);
            if (!ok) errs.Add(("出生日期", msg));
        }
        if (!string.IsNullOrEmpty(d["id_number"]))
        {
            var (ok, msg) = Validators.ValidateIdNumber(d["id_number"]);
            if (!ok) errs.Add(("身份证号", msg));
            else if (!string.IsNullOrEmpty(d["birth_date"]))
            {
                var (ok2, msg2) = Validators.ValidateBirthDateMatch(d["id_number"]!, d["birth_date"]!);
                if (!ok2) errs.Add(("出生日期/身份证号", msg2));
            }
            var dup = cn.QueryFirstOrDefault<long?>(
                "SELECT id FROM personnel_filing WHERE id_number=@n AND status='active'",
                new { n = d["id_number"] });
            if (dup is not null) errs.Add(("身份证号", "系统中已存在有效备案记录"));
        }
        return errs;
    }

    /// <summary>读取首个工作表，返回 (行号, 各列文本)，跳过表头行。</summary>
    private static List<(int RowNumber, List<string?> Cells)> ReadRows(Stream stream)
    {
        var result = new List<(int, List<string?>)>();
        using var doc = SpreadsheetDocument.Open(stream, false);
        var wbPart = doc.WorkbookPart!;
        var sheet = wbPart.Workbook.Descendants<Sheet>().FirstOrDefault();
        if (sheet is null) return result;
        var wsPart = (WorksheetPart)wbPart.GetPartById(sheet.Id!);
        var shared = wbPart.SharedStringTablePart?.SharedStringTable;

        foreach (var row in wsPart.Worksheet.Descendants<Row>())
        {
            var idx = (int)(row.RowIndex?.Value ?? 0);
            if (idx <= 1) continue;   // 跳过表头
            var cells = new List<string?>();
            var expected = 1;
            foreach (var c in row.Elements<Cell>())
            {
                // 补齐被跳过的空单元格，保证列位置对齐
                var colIdx = ColIndex(c.CellReference?.Value ?? "");
                while (expected < colIdx) { cells.Add(""); expected++; }
                cells.Add(CellText(c, shared));
                expected++;
            }
            result.Add((idx, cells));
        }
        return result;
    }

    private static string CellText(Cell c, SharedStringTable? shared)
    {
        if (c.DataType?.Value == CellValues.InlineString)
            return c.InlineString?.Text?.Text ?? "";
        var v = c.CellValue?.Text ?? "";
        if (c.DataType?.Value == CellValues.SharedString && shared is not null
            && int.TryParse(v, out var i) && i < shared.ChildElements.Count)
            return shared.ChildElements[i].InnerText;
        return v;
    }

    private static int ColIndex(string reference)
    {
        var n = 0;
        foreach (var ch in reference)
        {
            if (!char.IsAsciiLetter(ch)) break;
            n = n * 26 + (char.ToUpperInvariant(ch) - 'A' + 1);
        }
        return n == 0 ? 1 : n;
    }

    /// <summary>生成导入模板（表头 + 一行示例 + 说明页）。</summary>
    public static byte[] GenerateTemplate(IDbConnection cn)
    {
        var sample = new List<string?[]>
        {
            new string?[]
            {
                "总部", "人事处", "张三", "男", "19900101", "20120701", "110101199001012133",
                "北京市朝阳区", "群众", "科长", "人事处", "03", "03", "03", "03", "", "科长", "新增", "是", "",
            },
        };
        string Codes(string cat) => string.Join("；",
            Helpers.GetDictOptions(cn, cat).Select(o => $"{o.Code}={o.Value}"));

        using var w = new ExcelWriter();
        w.AddSheet("导入模板", "备案人员批量导入模板",
            Columns.Select(c => c.Header).ToArray(), sample, notes:
            [
                "1. 带 * 的列为必填项；第 2 行为示例数据，导入前请删除。",
                "2. 日期一律填 8 位数字 YYYYMMDD，如 19900101。",
                "3. 出生日期须与身份证号第 7–14 位一致；身份证号须通过校验位算法。",
                "4. 系统中已存在同身份证号的有效备案记录时，该行会被拒绝并在报告中列出。",
                "5. 操作人由当前登录账户自动记录，无需在表格中填写。",
                $"6. 学历代码：{Codes("education")}",
                $"7. 学位代码：{Codes("degree")}",
                $"8. 职称代码：{Codes("title")}",
                $"9. 职级代码：{Codes("rank")}",
            ]);
        return w.ToArray();
    }
}
